src/zones/erldns_zone_cache.erl

```diff
@@ -626,29 +664,136 @@ sign_zone(Zone) ->
+% looking for both exact and wildcard matches.
+record_name_in_zone_with_wildcard_strict(_, []) ->
+    [];
+record_name_in_zone_with_wildcard_strict(ZoneLabels, [_ | _] = RecordLabels) ->
+    Parent = lists:droplast(RecordLabels),
+    WildcardLabels = Parent ++ [~"*"],
+    case pattern_zone_dname(ZoneLabels, WildcardLabels) of
+        [] ->
+            case pattern_zone_dname(ZoneLabels, Parent) of
+                [] ->
+                    record_name_in_zone_with_wildcard_strict(ZoneLabels, Parent);
+                RRsStrict ->
+                    RRsStrict
+            end;
+        RRsWild ->
+            RRsWild
+    end.
+
+is_record_name_in_zone_helper(ZoneLabels, RecordLabels) ->
+    is_record_name_in_zone_traverse_wildcard(ZoneLabels, RecordLabels, RecordLabels).
+
+is_record_name_in_zone_with_wildcard(_, []) ->
+    false;
+is_record_name_in_zone_with_wildcard(ZoneLabels, QLabels) ->
+    Parent = lists:droplast(QLabels),
+    WildcardPath = Parent ++ [~"*"],
+    is_record_name_in_zone_traverse_wildcard(ZoneLabels, WildcardPath, Parent).
+
+is_record_name_in_zone_traverse_wildcard(ZoneLabels, Path, ParentPath) ->
+    0 =/= pattern_zone_dname_count(ZoneLabels, Path) orelse
+        is_record_name_in_zone_with_wildcard(ZoneLabels, ParentPath).
+
+is_record_name_in_zone_with_descendants(ZoneLabels, QLabels) ->
+    % eqwalizer:ignore this needs to be an improper list for tree traversal
+    HasDescendantsPath = QLabels ++ '_',
+    0 =/= pattern_zone_dname_count(ZoneLabels, HasDescendantsPath).
+
+pattern_zone(ZoneLabels) ->
+    Pattern = {{{ZoneLabels, '_', '_'}, '$1'}, [], ['$1']},
+    lists:append(ets:select(erldns_zone_records_typed, [Pattern])).
+
+pattern_zone_dname(ZoneLabels, Labels) ->
+    Pattern = {{{ZoneLabels, Labels, '_'}, '$1'}, [], ['$1']},
+    lists:append(ets:select(erldns_zone_records_typed, [Pattern])).
+
+pattern_zone_dname_type(ZoneLabels, Labels, Type) ->
+    Pattern = {{{ZoneLabels, Labels, Type}, '$1'}, [], ['$1']},
+    lists:append(ets:select(erldns_zone_records_typed, [Pattern])).
+
+pattern_zone_dname_count(ZoneLabels, Labels) ->
+    Pattern = {{{ZoneLabels, Labels, '_'}, '$1'}, [], [true]},
+    ets:select_count(erldns_zone_records_typed, [Pattern]).
+
+pattern_zone_dname_type_delete(ZoneLabels, Labels, Type) ->
+    Pattern = {{{ZoneLabels, Labels, Type}, '_'}, [], [true]},
+    ets:select_delete(erldns_zone_records_typed, [Pattern]).
+
+pattern_zone_delete(ZoneLabels) ->
+    Pattern = {{{ZoneLabels, '_', '_'}, '_'}, [], [true]},
+    ets:select_delete(erldns_zone_records_typed, [Pattern]).
+
+fix_tables(false, _) ->
+    ets:safe_fixtable(erldns_zone_records_typed, false),
+    ets:safe_fixtable(erldns_zones_table, false);
+fix_tables(true, ZoneLabels) ->
+    case lookup_zone(ZoneLabels) of
+        zone_not_found ->
+            ok;
+        #zone{} ->
+            ets:safe_fixtable(erldns_zone_records_typed, true),
+            ets:safe_fixtable(erldns_zones_table, true)
+    end.
```

Comment:
This is pretty interesting. So by fixing the table during a zone update we avoid the case of missing zone records for answer synthesis especially for signed zones, or am I misunderstanding the utility?

Reply:
Actually, I might have gotten this one wrong 🤦🏽 This would indeed be the case if we were doing table traversals using `select_match` or `next`, but all operations we do on the tables are just `lookup`s and `select`s, so it makes no difference except for affecting concurrency performance 🤦🏽 

Anyway, this would only happen if you were `put`ing a zone that already existed, which you could entirely avoid if you'd do first delete on the old records and then insert on the new ones instead. And it's a bit of an extreme edge case, the delete-then-insert is so fast.